core/lib/spree/event.rb

```diff
@@ -6,44 +6,102 @@
+  #
+  # Then, you can use {#subscribe} to hook into the event:
+  #
+  #   @example
+  #     Spree::Event.subscribe 'order_finalized' do |event|
+  #       # Take the order at event.payload[:order]
+  #     end
+  #
+  # You can also subscribe to an event through a module including
+  # {Spree::Event::Subscriber}:
+  #
+  #  @example
+  #    module MySubscriber
+  #      include Spree::Event::Subscriber
+  #
+  #      event_action :order_finalized
+  #
+  #      def order_finalized(event)
+  #        # Again, take the order at event.payload[:order]
+  #      end
+  #    end
   module Event
     extend self
 
     delegate :activate_autoloadable_subscribers, :activate_all_subscribers, :deactivate_all_subscribers, to: :subscriber_registry
 
-    # Allows to trigger events that can be subscribed using #subscribe. An
-    # optional block can be passed that will be executed immediately. The
-    # actual code implementation is delegated to the adapter.
+    # Allows to trigger events that can be subscribed using {#subscribe}.
     #
-    # @param [String] event_name the name of the event. The suffix ".spree"
-    #  will be added automatically if not present
-    # @param [Hash] opts a list of options to be passed to the triggered event
+    # The actual code implementation is delegated to the adapter.
+    #
+    # @param [String, Symbol] event_name the name of the event.
+    # @param [Hash] opts a list of options to be passed to the triggered event.
+    # They will be made available through the {Spree::Event::Event} instance
+    # that is yielded to the subscribers (see {Spree::Event::Event#payload}).
+    # However, take into account that the deprecated
+    # {Spree::Event::Adapters::ActiveSupportNotifications} adapter will yield a
+    # {ActiveSupport::Notifications::Fanout::Subscribers::Timed} instance
+    # instead.
+    # @option opts [Any] :adapter Reserved to indicate the adapter to use as
+    # event bus. Defaults to {#default_adapter}
     #
     # @example Trigger an event named 'order_finalized'
     #   Spree::Event.fire 'order_finalized', order: @order do
     #     @order.finalize!
     #   end
+    #
+    # TODO: Change opts to be keyword arguments and include `adapter:` in them
     def fire(event_name, opts = {})
-      adapter.fire normalize_name(event_name), opts do
-        yield opts if block_given?
+      adapter = opts.delete(:adapter) || default_adapter
+      if block_given? && !legacy_adapter(adapter)
+        raise ArgumentError, <<~MSG
+          Blocks passed to `Spree::Event.fire` are ignored unless the adapter is
+          `Spree::Event::Adapters::ActiveSupportNotifications` (which is
+          deprecated). Please, instead of:
+
+            Spree::Event.fire 'event_name', order: order do
+              order.do_something
+            end
+
+          Use:
+
+            order.do_something
+            Spree::Event.fire 'event_name', order: order
+
+        MSG
+      elsif block_given?
+        yield opts
```

Comment:
I think we need to emit a deprecation warning here, otherwise people using `AS::N` adapter won't see that they need to remove the block in their fire calls. They will only know this with an exception after switching to the new adapter when `fire` will be called.

Reply:
The main deprecation message, the one they should see when booting up the app or running the test suite, is already informing users to stop providing blocks to `Spree::Event.fire`. Do you think it's not enough? In fact, the block is still supported if they're using the legacy adapter.

Reply:
I saw that the main deprecation warning mentions that people need to stop using fire with block, but that code could technically be in some extension and they won't be able to know it until the code breaks. Honestly, I don't think it's super used in the real world but I'd still add a deprecation warning here, just to be 100% sure.

Reply:
I see. Added!